src/Microsoft.AspNet.Server.Kestrel/Infrastructure/ThreadPoolActions.cs

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Server.Kestrel.Http;
+
+namespace Microsoft.AspNet.Server.Kestrel.Infrastructure
+{
+    public class ThreadPoolActions
+    {
+        private readonly IKestrelTrace _log;
+
+        private readonly WaitCallback _runAction;
+        private readonly WaitCallback _completeTcs;
+        private readonly WaitCallback _abortFrame;
+        private readonly WaitCallback _returnBlocks;
+        private readonly WaitCallback _returnWrittenBlocks;
+
+        public ThreadPoolActions(IKestrelTrace log)
+        {
+            _log = log;
+
+            // Curry and capture log in closures once
+            _runAction = (o) =>
+            {
+                try
+                {
+                    ((Action)o)();
+                }
+                catch (Exception e)
+                {
+                    _log.ApplicationError(e);
+                }
+            };
+
+            _completeTcs = (o) =>
+            {
+                try
+                {
+                    ((TaskCompletionSource<object>)o).SetResult(null);
```

Comment:
(For reference it ends up in this code, eventually, on SetResult https://github.com/dotnet/coreclr/blob/29526273e376bdf647102388260907b70778f9df/src/mscorlib/src/System/Threading/Tasks/Task.cs#L3581)